### PR TITLE
Fix/Re-implement Bug Bite

### PR DESCRIPTION
**Pull request by ghoulslash**
- new bug bite implementation that does not temporarily change items
- Addresses #2027 
- Fixes issue where bug bite was not activating if it fainted the target
- Save original gBattlerTarget through the move effect to fix `tryfaintmon BS_TARGET` following the effect
- Jaboca/etc berries still need to be activated prior to bug bite stealing them, but that should be saved for another PR

bugbite

as noted [here](https://discord.com/channels/419213663107416084/419214240277200898/1011470451861635113), GetBattlerHoldEffectParam is used to determine certain berry effects, but since this PR does not temporarily swap items, we can't use that function.

### CODE REVIEW

**Review by AsparagusEduardo — COMMENTED** [review]
I see that some `TRUE`s and `FALSE`s where changed to 1 and 0. Those should probably need to be reverted.

**Comment by LOuroboros** [comment]
Is there a reason not to use `TRUE`/`FALSE` for `sBERRY_OVERRIDE`? :eyes:

**Comment by ghoulslash** [comment]
> Is there a reason not to use `TRUE`/`FALSE` for `sBERRY_OVERRIDE`? 👀

because I changed it to be a "state", see: <https://github.com/rh-hideout/pokeemerald-expansion/pull/2138/files#diff-658ef5bc9368700b8dbd6f4137cf50a00220f0eb79e352bb9858047c45a4e09bR7125>

File: src/battle_util.c [hunk 1]
```diff
@@ -6935,189 +7120,24 @@ u8 ItemBattleEffects(u8 caseID, u8 battlerId, bool8 moveTurn)
             }
         }
         break;
-    case ITEMEFFECT_BATTLER_MOVE_END:
-        goto DO_ITEMEFFECT_MOVE_END;    // this hurts a bit to do, but is an easy solution
-    case ITEMEFFECT_MOVE_END:
-        for (battlerId = 0; battlerId < gBattlersCount; battlerId++)
+    case ITEMEFFECT_USE_LAST_ITEM:
+        effect = ItemEffectMoveEnd(battlerId, ItemId_GetHoldEffect(gLastUsedItem));
+        gBattleScripting.overrideBerryRequirements = 2; // to exit VARIOUS_CONSUME_BERRIES
```
Comment (LOuroboros) [inline src/battle_util.c:7125]:
```suggestion
        gBattleScripting.overrideBerryRequirements = 2; // to exit VARIOUS_CONSUME_BERRY
```

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]
Jaboca Berry's effect isn't activating before Bug Bite.
mGBA_ZPRI5iQ7zg

**Comment by StephenLynx** [comment]
I tested it out on IE and it seems to work fine.

**Comment by StephenLynx** [comment]
I found the issue on my port: in battle_util GetBattlerHoldEffectParam was used with the id of the attacker while itemId was ignored in a few functions. So I used ItemId_GetHoldEffectParam directly with itemId in these places. this PR probably carries the same issue. https://gitgud.io/stephenlynx/pokemonie/-/commit/02107204b6f03d1f0d7a11cf69ec5d56ef40ea45

**Comment by StephenLynx** [comment]
I fixed enigma berries, by the way. I was not sure if my commit broke them, because I couldn't find how the table was populated, but if they were broken, now they are surely fixed. 
https://gitgud.io/stephenlynx/pokemonie/-/commit/c0eb452d0ffbe43e669ea8ff0937bbaec6526b5e

**Comment by StephenLynx** [comment]
I do believe to have found another bug. When it checks if the case is ITEMEFFECT_MOVE_END, there is a loop through the mons in the battle. Later checks that make use of gLastUsedItem will not be updated for subsequential mons, since it was set the last time right after ItemBattleEffects started on the check caseID != ITEMEFFECT_BATTLER_MOVE_END. I fixed it by setting gLastUsedItem to the item of the current on on the loop. 
https://gitgud.io/stephenlynx/pokemonie/-/commit/a2be0a20aebb1ed7051ef43b24632fb80dfd6d2c

**Pull request closed without merging**